src/detect-dns-query.c

```diff
@@ -612,336 +612,9 @@ static int DetectDnsQueryTest03(void)
-    f.alproto = ALPROTO_DNS;
-
-    p1->flow = &f;
-    p1->flags |= PKT_HAS_FLOW|PKT_STREAM_EST;
-    p1->flowflags |= FLOW_PKT_TOSERVER|FLOW_PKT_ESTABLISHED;
-
-    p2->flow = &f;
-    p2->flags |= PKT_HAS_FLOW|PKT_STREAM_EST;
-    p2->flowflags |= FLOW_PKT_TOSERVER|FLOW_PKT_ESTABLISHED;
-
-    StreamTcpInitConfig(TRUE);
-
-    DetectEngineCtx *de_ctx = DetectEngineCtxInit();
-    FAIL_IF_NULL(de_ctx);
-    de_ctx->mpm_matcher = mpm_default_matcher;
-    de_ctx->flags |= DE_QUIET;
-
-    s = DetectEngineAppendSig(de_ctx, "alert dns any any -> any any "
-                              "(msg:\"Test dns_query option\"; "
-                              "dns_query; content:\"google\"; nocase; sid:1;)");
-    FAIL_IF_NULL(s);
-
-    SigGroupBuild(de_ctx);
-    DetectEngineThreadCtxInit(&tv, (void *)de_ctx, (void *)&det_ctx);
-
-    FLOWLOCK_WRLOCK(&f);
-    int r = AppLayerParserParse(NULL, alp_tctx, &f, ALPROTO_DNS,
-                                STREAM_TOSERVER, buf1, sizeof(buf1));
-    if (r != 0) {
-        printf("toserver chunk 1 returned %" PRId32 ", expected 0: ", r);
-        FLOWLOCK_UNLOCK(&f);
-        FAIL;
-    }
-    FLOWLOCK_UNLOCK(&f);
-
-    dns_state = f.alstate;
-    FAIL_IF_NULL(dns_state);
-
-    /* do detect */
-    SigMatchSignatures(&tv, de_ctx, det_ctx, p1);
-
-    if (PacketAlertCheck(p1, 1)) {
-        printf("sig 1 alerted, but it should not have: ");
-        FAIL;
-    }
-
-    FLOWLOCK_WRLOCK(&f);
-    r = AppLayerParserParse(NULL, alp_tctx, &f, ALPROTO_DNS, STREAM_TOSERVER,
-                            buf2, sizeof(buf2));
-    if (r != 0) {
-        printf("toserver chunk 1 returned %" PRId32 ", expected 0\n", r);
-        FLOWLOCK_UNLOCK(&f);
-        FAIL;
-    }
-    FLOWLOCK_UNLOCK(&f);
-
-    /* do detect */
-    SigMatchSignatures(&tv, de_ctx, det_ctx, p2);
-
-    if (!(PacketAlertCheck(p2, 1))) {
-        printf("sig 1 didn't alert, but it should have: ");
-        FAIL;
-    }
-
-    if (alp_tctx != NULL)
-        AppLayerParserThreadCtxFree(alp_tctx);
-    if (det_ctx != NULL)
-        DetectEngineThreadCtxDeinit(&tv, det_ctx);
-    if (de_ctx != NULL)
-        SigGroupCleanup(de_ctx);
-    if (de_ctx != NULL)
-        DetectEngineCtxFree(de_ctx);
-
-    StreamTcpFreeConfig(TRUE);
-    FLOW_DESTROY(&f);
-    UTHFreePacket(p1);
-    UTHFreePacket(p2);
-    PASS;
-}
-
-/** \test simple google.com query matching (TCP splicing) */
-static int DetectDnsQueryTest05(void)
-{
-    /* google.com in 2 chunks (buf1 and buf2) */
-    uint8_t buf1[] = {  0x00, 28,                               /* len 28 */
-                        0x10, 0x32, 0x01, 0x00, 0x00, 0x01,
-                        0x00, 0x00, 0x00, 0x00, 0x00, 0x00, };
-
-    uint8_t buf2[] = {  0x06, 0x67, 0x6F, 0x6F, 0x67, 0x6C,
-                        0x65, 0x03, 0x63, 0x6F, 0x6D, 0x00,
-                        0x00, 0x10, 0x00, 0x01, };
-
-    uint8_t buf3[] = {  0x00, 44,                               /* len 44 */
```

Comment:
This test did not get translated to Rust nor Suricata-verify, did it ?

You can take a look at https://github.com/OISF/libhtp/pull/269 to produce a pcap out of these buffers

Reply:
Is it necessary to translate the full test to Rust? Makes sense as a Suricata-Verify test I guess, but out of scope for a DNS unit test.

Reply:
Worth considering for a new protos. Do they require this test? In theory if Suricata gets it right for one proto, it chould get it right for all, provided the proto returns the right response code - which can be tested as a unit test.

Because creating S-V tests for this is much harder if you don't already have a pcap ready.

Reply:
> Because creating S-V tests for this is much harder if you don't already have a pcap ready.

Suricata-Verify is ok indeed, cf the proposed PR to create a pcap out of this unit test

> In theory if Suricata gets it right for one proto, it should get it right for all, provided the proto returns the right response code

Well, that kind of mistake happens cf https://github.com/OISF/suricata/pull/4980

Indeed, we can reduce this to a unit test that `Incomplete` gets returned with the right values

Reply:
> Well, that kind of mistake happens cf #4980

Can you think of an easy way to create SV tests for this? I guess it would need to take a pcap, re organize the data over more TCP packets..  Not saying it shouldn't need to be tested, but how we can go about it.

Reply:
> Can you think of an easy way to create SV tests for this?

Yes, I have been mentioning OISF/libhtp#269 which does this kind of thing, ie generate a pcap from some buffers input
Actually, it makes a client and server communicate over TCP, and you can record it as a pcap.
I have been modifying this script to use with other stuff than HTTP, using one line with hexadecimal encoding per packet to send.

I do not know if it is easy enough. And the downside is that this is an artificial pcap, not a regular software communicating...

Reply:
So what do you need provide? Just provide some payload buffers?

Maybe we need to keep tests like the DetectDnsQueryTest05 that was removed, but a simpler way to bootstrap them.

Reply:
The input file should be something like
```
s2c 650342
c2s 1010
c2s 1212121516
s2c 667EF4CC
``` 

Do you want me to tweak the htptopcap.py script to use an input file like this ?

Reply:
@catenacyber Please do and we can give it a try?

Reply:
Here is a python script and its input file.

txt2pcap.py
```
import sys
import binascii
from threading import Thread
import time
import socket

# Create a pcap from a htp test file
# Launches a server on port 8080
# Launches a client in another thread that connects to it
# Both client and server read the htp test file
# And they send and receive data as described (without analysing it)
# So, you need to capture traffic on port 8080 while running the script

def removeOneEOL(s):
    r = s
    if r[-1] == '\n':
        r = r[:-1]
        if r[-1] == '\r':
            r = r[:-1]
    return r

PCAP_TCP_PORT = 5353

class ServerThread(Thread):

    def __init__(self, filename):
        Thread.__init__(self)
        self.filename = filename

    def run(self):
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.bind(("127.0.0.1", PCAP_TCP_PORT))
        s.listen(1)
        conn, addr = s.accept()
        f = open(self.filename)
        sending = ""
        receiving = ""

        for l in f.readlines():
            data = binascii.unhexlify(l.split()[1])
            if l.split()[0] == "s2c":
                conn.send(data)
                print "server sent", len(data)
            else:
                data = conn.recv(len(data))
                print "server recvd", len(data)

        conn.close()
        s.close()
        f.close()


class ClientThread(Thread):

    def __init__(self, filename):
        Thread.__init__(self)
        self.filename = filename

    def run(self):
        time.sleep(1)
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.connect(("127.0.0.1", PCAP_TCP_PORT))
        f = open(self.filename)
        sending = ""
        receiving = ""

        for l in f.readlines():
            data = binascii.unhexlify(l.split()[1])
            if l.split()[0] != "s2c":
                s.send(data)
                print "client sent", len(data)
            else:
                data = s.recv(len(data))
                print "client recvd", len(data)

        s.close()
        f.close()

t1 = ServerThread(sys.argv[1])
t2 = ClientThread(sys.argv[1])

# Launch threads
t1.start()
t2.start()

# Wait for threads to finish
t1.join()
t2.join()
```

dns.txt
```
c2s 001c103201000001000000000000
c2s 06676F6F676C6503636F6D0000100001
c2s 002c10328180000100010000000006676F6F676C6503636F6D0000010001c00c00010001000140ef000401020304
c2s 001c11330100000100000000000006676F6F676C65036E65740000100001
```

Run `python txttopcap.py dns.txt` while capturing traffic and you get a nice pcap
Wireshark shoed the tcp reassembly nicely for the 2 first packets

Reply:
Whats the status of this PR?

Reply:
This PR is OK as is. The question is more generic in how should any protocol that uses the engine provided buffering be tested?